### ISSUE TITLE
CA2007 - Don't warn on `Task.CompletedTask` & `Task.From*()`

### ISSUE THREAD

**Issue by antmdvs** [dotnet/roslyn-analyzers#3261]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

Example: [CA2007](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2007-do-not-directly-await-task)

#### Repro steps

```csharp
await Task.CompletedTask;
```
or
```csharp
await Task.FromResult(x);
```
#### Expected behavior

There should be no need for `ConfigureAwait` since these tasks are already completed so no delegate will run. I believe the same is true for all `Task.From*` methods.

#### Actual behavior

Analyzer warns

**Comment by antmdvs** [comment]
It may be the case that the expressions in question are more likely to be `return`ed than `await`ed. If that is the recommended usage pattern, then the warning should probably say something to that effect instead.

**Comment by sharwell** [comment]
📝 CA2007 has been superseded by [VSTHRD111](https://github.com/microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD111.md)

**Comment by Evangelink** [comment]
@sharwell Probably worth moving this ticket (if you have the right to do so).

**Comment by Evangelink** [comment]
@mavasani @sharwell My understanding is that this ticket should be moved to the other repo, is that right? If so do you have the right to do the move?

**Comment by Evangelink** [comment]
@mavasani Based on @AArnott comment's on another thread I think this ticket shall be marked as `Blocked` while the `VSTHRD111` is moved back to this repo.

**Comment by AArnott** [comment]
VSTHRD111 won't be moved. Either CA2007 or a successor will be what we use going forward. So this isn't blocked -- it could be fixed now.
But FWIW, I would personally resolve this as Won't Fix, since *awaiting* a `Task.CompletedTask;` or `Task.FromResult` call is a complete waste of characters and the await should be removed, which would remove the warning as well. Or am I missing something?

**Comment by Evangelink** [comment]
> 
> VSTHRD111 won't be moved. Either CA2007 or a successor will be what we use going forward. So this isn't blocked -- it could be fixed now.

I guess I didn't get what's going-on with the move of the analyzers but nevermind.

>
> But FWIW, I would personally resolve this as Won't Fix, since _awaiting_ a `Task.CompletedTask;` or `Task.FromResult` call is a complete waste of characters and the await should be removed, which would remove the warning as well. Or am I missing something?

I agree that awaiting such task is pointless but the warning is talking about potential deadlock and threading issue which isn't the case for such task. So I would say that having an exclusion would be nice and we could think of having another rule or updating an existing rule to report a different issue to say this is useless. For example we could use #3388.

**Comment by mavasani** [comment]
> I would personally resolve this as Won't Fix

Aye.
